lib/Core/Executor.cpp

```diff
@@ -3555,7 +3555,37 @@ void Executor::executeMemoryOperation(ExecutionState &state,
     if (MaxSymArraySize && mo->size >= MaxSymArraySize) {
       address = toConstant(state, address, "max-sym-array-size");
     }
-    
+    // handle concrete address case first
+    if (auto *c_address = dyn_cast<ConstantExpr>(address)) {
+      auto first_offset = mo->getOffset(c_address->getZExtValue());
+      auto first_out_of_bound = mo->getBoundsCheckOffset(bytes);
+
+      // handle out-of-bounds access
+      if (first_offset >= first_out_of_bound) {
+        terminateStateOnError(state, "memory error: out of bound pointer", Ptr,
+                              NULL, getAddressInfo(state, address));
+        return;
+      }
+      const ObjectState *os = op.second;
+      if (isWrite) {
+        if (os->readOnly) {
+          terminateStateOnError(state, "memory error: object read only",
+                                ReadOnly);
+        } else {
+          ObjectState *wos = state.addressSpace.getWriteable(mo, os);
+          wos->write(first_offset, value);
+        }
+      } else {
+        ref<Expr> result = os->read(first_offset, type);
+
+        if (interpreterOpts.MakeConcreteSymbolic)
+          result = replaceReadWithSymbolic(state, result);
+
+        bindLocal(target, state, result);
+      }
+
+      return;
+    }
```

Comment:
I don't really like this code, it basically duplicates the non concrete case and just makes everything more complex.  I don't think `klee` should double in complexity,  just to keep interpretation overhead in line with `lli`, however if you want that, I think you should template it somehow, the `auto` might actually be useful in a templated version,

Reply:
Well, KLEE does support a mix of concrete and symbolic values. And typically, when you execute your application, a lot of it is not symbolic but concrete. So, optimising towards having `lli` as a baseline is not that bad.

Reply:
Of course it's not bad, the question is what trade off are you willing to make.

What you are doing in parts of this PR is essentially duplicating functionality and optimising for the concrete case. I think if you take this approach to its conclusion, you would double the interpretation code size, with one half being current KLEE and the other part being `lli`. While this would make KLEE a faster interpreter, it would in my opinion make it a worse tool. It would be harder to maintain, extend and onboard new people.

Reply:
Nope, I'm handling hot cases. And this is one. 50% reduction of the original KLEE is not just a corner case. And it's not even close to `lli`.

Reply:
I'm talking about a "simple" refactoring, what I'm proposing should not change the compiled binary. I'm not sure what you are talking about.

Reply:
Now i finally understand what your issue is:
```
auto first_offset = mo->getOffset(c_address->getZExtValue());
```
`first_offset` is a `uint64_t` so it uses the concrete functions thanks to function overloading.
So, it can't be a "simple" refactoring ;)

Reply:
Yes, but `uint64_t` and `ref<Expr>` behave veryyyy similarly in this case, so you could template it.

Reply:
of course, that's the power of abstraction, but one (`uint64_t`) is super fast stack allocation, the other one involves one heap allocation that gets destroyed almost immediately later.